nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -213,93 +215,157 @@ impl DataStore {
+                    bail!(err_msg);
                 }
             }
         }
 
         if !all_versions.contains(&current_version) {
-            return Err(format!(
+            bail!(
                 "Current DB version {current_version} was not found in {}",
                 config.schema_dir
-            ));
+            );
         }
         // TODO: Test this?
         if !all_versions.contains(&desired_version) {
-            return Err(format!(
+            bail!(
                 "Target DB version {desired_version} was not found in {}",
                 config.schema_dir
-            ));
+            );
         }
 
         let target_versions = all_versions.range((
             Bound::Excluded(&current_version),
             Bound::Included(&desired_version),
         ));
 
-        for target_version in target_versions.into_iter() {
+        // Iterate over semver versions (e.g., "1.0.0 -> 2.0.0").
+        for mut target_version in target_versions.into_iter().map(|v| v.clone())
+        {
             info!(
                 log,
                 "Attempting to upgrade schema";
-                "current_version" => current_version.to_string(),
-                "target_version" => target_version.to_string(),
+                "current_version" => %current_version,
+                "target_version" => %target_version,
             );
 
             let target_dir = config.schema_dir.join(target_version.to_string());
 
-            let schema_change =
-                all_sql_for_version_migration(&target_dir).await?;
+            let schema_change = all_sql_for_version_migration(&target_dir)
+                .await
+                .map_err(anyhow::Error::msg)
+                .context("Failed to access all SQL statements for a version migration")?;
 
-            // Confirm the current version, set the "target_version"
-            // column to indicate that a schema update is in-progress.
+            // Iterate over each individual file that comprises a schema change.
             //
-            // Sets the following:
-            // - db_metadata.target_version = new version
-            self.prepare_schema_update(&current_version, &target_version)
+            // For example, this contains "up01.sql", "up02.sql", etc.
+            //
+            // While this update happens, the "current_version" will remain
+            // the same (unless another Nexus concurrently completes the
+            // update), but the "target_version" will keep shifting on each
+            // incremental step.
+            let mut prior_step_version = None;
+            for (i, SchemaUpgradeStep { path: _, sql }) in
+                schema_change.steps.iter().enumerate()
+            {
+                // Durably store our progress in the target version as the
+                // pre-release extension of "target_version".
+                target_version.0.pre =
+                    semver::Prerelease::new(&format!("step.{i}"))
```

Comment:
It might be worth a comment that the `.` here is load-bearing (I had initially suggested `step-{i}`, which would be incorrect). From https://docs.rs/semver/1.0.22/semver/struct.Prerelease.html#examples:

> Be aware that if you are planning to number your own pre-releases, you should prefer to separate the numeric part from any non-numeric identifiers by using a dot in between. That is, prefer pre-releases alpha.1, alpha.2, etc rather than alpha1, alpha2 etc. The SemVer spec’s rule for pre-release precedence has special treatment of numeric components in the pre-release string, but only if there are no non-digit characters in the same dot-separated component. So you’d have alpha.2 < alpha.11 as intended, but alpha11 < alpha2.